pipeline/dags/dag_utils/sources/monenfant.py

```diff
@@ -1,12 +1,193 @@
+import contextlib
 import json
 import logging
+import os
+import re
+from datetime import datetime
 from pathlib import Path
+from typing import Literal
+
+import httpx
+from playwright.sync_api import (
+    Browser,
+    TimeoutError as PlaywrightTimeoutError,
+    sync_playwright,
+)
+from twocaptcha import TwoCaptcha
 
 from . import utils
 
 logger = logging.getLogger(__name__)
 
 
+def get_location(commune: str, region: str) -> str:
+    """Return a location string suitable for the search form on monenfant.fr"""
+    match_paris = re.search(r"Paris (\d+)", commune)
+    if match_paris is not None:
+        num_arrondissement = match_paris[1]
+        suffix = "er" if num_arrondissement == "1" else "eme"
+        return f"{num_arrondissement}{suffix} Arrondissement Paris"
+
+    return f"{commune} {region}"
+
+
+@contextlib.contextmanager
+def solve_captcha(page, timeout: int = 500):
+    """Wait for captcha and solve it if ever"""
+
+    # 1. wait for captcha to show up if ever
+    try:
+        with page.expect_response("**/captcha", timeout=timeout) as response_info:
+            # here, calling code should do actions expected to trigger captcha
+            yield
+    except PlaywrightTimeoutError:
+        # captcha did not show up : simply ignore
+        return
+
+    response = response_info.value
+    data = response.json()
+
+    TWOCAPTCHA_API_KEY = os.environ.get("TWOCAPTCHA_API_KEY")
+    captcha_solver = TwoCaptcha(apiKey=TWOCAPTCHA_API_KEY)
+
+    # 2. try to solve captcha challenges
+    for _ in range(5):
```

Comment:
pourquoi 5 ? on peut augmenter à l'usage si ça foire ?

Reply:
le captcha solver fonctionne très bien sur les captchas proposés sur le site monenfant.fr. Je dirais 9 fois sur 10 ça marche. Donc très peu probable de rater plus de 2 fois.

Cela dit j'imagine qu'on n'est pas à l'abri d'une régression sur le service 2captcha ou sur un changement de captcha sur monenfant... étant donné que le service 2captcha est en mode pay-as-you-go, la limite permet de s'assurer qu'on ne crame pas tout le crédit dans une boucle :sweat_smile: 

(commentaire ajouté)